File: server/dotnet/sdk-dotnet-example/Business/Handlers/CreditCardPaymentHandler.cs
using Business.Domain.Enums.Payments;
using Business.DTOs.CreatePayments;
using Business.DTOs.Services;
using Business.Interfaces.Handlers;
using Business.Interfaces.SDKClients;

namespace Business.Handlers;

public class CreditCardPaymentHandler(IPaymentClient paymentClient, IServiceClient serviceClient) : IPaymentMethodHandler
{
    public PaymentMethodType SupportedMethod => PaymentMethodType.CREDIT_CARD;

    public async Task<CreatePaymentResponseDto> HandleAsync(CreatePaymentRequestDto request)
    {
        if (!string.IsNullOrEmpty(request.Card?.Number))
        { 
            GetIinDetailsResponseDto? responseDto = await serviceClient.GetIinDetails(new()
            {
                Bin = request.Card.Number,
            });
            request.PaymentProductId = responseDto?.PaymentProductId;
        }
        
        return await paymentClient.CreatePaymentAsync(request);
    }
}

File: server/ruby/app/business/handlers/credit_card_payment_handler.rb
module Business
  module Handlers
    class CreditCardPaymentHandler
      def initialize(payment_client:, service_client:)
        @payment_client = payment_client
        @service_client = service_client
      end

      def supported_method
        Business::Domain::Enums::Payments::PaymentMethodType::CREDIT_CARD
      end

      def handle(request_dto)

        get_payment_product_id_request = Business::Dtos::Service::GetPaymentProductId::RequestDto.new(
          request_dto.card.number
        )

        get_payment_product_id_response = @service_client.get_payment_product_id(get_payment_product_id_request)

        request_dto.payment_product_id = get_payment_product_id_response.payment_product_id

        @payment_client.create_payment(request_dto)
      end
    end
  end
end
